File: ios/Services/YandexMusic/Sources/YandexMusicService/Parsing/YandexMusicMappers.swift
import CoreDomain
import Foundation

// MARK: - Account status

extension [String: Any] {
    func toAvailability() throws -> MusicServiceAvailability {
        let account = try requiredObject("account")
        let permissions = optionalObject("permissions")
        let permissionValues = permissions?
            .optionalAnyArray("values")?
            .compactMap { $0 as? String } ?? []
        return MusicServiceAvailability(
            isAvailable: try account.requiredBool("serviceAvailable"),
            region: account.optionalInt64("region").map(Int.init),
            permissions: Set(permissionValues)
        )
    }

    func toCurrentUserId() throws -> ProviderUserId {
        guard let id = optionalStringOrInt("id") ?? optionalStringOrInt("default_uid") else {
            throw MusicLibraryError.invalidResponse(
                description: "Yandex user info response does not contain user id."
            )
        }
        return ProviderUserId(rawValue: id)
    }
}

// MARK: - Playlists

extension [String: Any] {
    func toPlaylistSummary() throws -> PlaylistSummary {
        let owner = try requiredObject("owner")
        let ownerUid = owner.optionalInt64("uid").map(String.init)
        let ownerId = try ownerUid ?? String(requiredInt64("uid"))
        return PlaylistSummary(
            id: PlaylistId(
                ownerId: ProviderUserId(rawValue: ownerId),
                kind: PlaylistKind(rawValue: try requiredInt64("kind"))
            ),
            provider: .yandexMusic,
            playlistUuid: optionalString("playlistUuid").map(PlaylistUuid.init(rawValue:)),
            title: try requiredString("title"),
            ownerName: owner.optionalString("name"),
            coverUriTemplate: coverUriTemplate(),
            trackCount: Int(try requiredInt64("trackCount")),
            durationMs: optionalInt64("durationMs"),
            isAvailable: optionalBool("available") ?? true,
            isCollective: optionalBool("collective") ?? false,
            visibility: optionalString("visibility").flatMap(\.toPlaylistVisibility),
            role: playlistRole()
        )
    }

    func toPlaylist() throws -> Playlist {
        let trackItems = optionalObjectArray("tracks") ?? []
        let tracks = try trackItems.enumerated().map { index, element in
            try element.toPlaylistTrackEntry(position: index)
        }
        return Playlist(
            summary: try toPlaylistSummary(),
            revision: optionalInt64("revision"),
            snapshot: optionalInt64("snapshot"),
            likesCount: optionalInt64("likesCount").map(Int.init),
            tracks: tracks
        )
    }

    private func toPlaylistTrackEntry(position: Int) throws -> PlaylistTrackEntry {
        let track = try optionalObject("track")?.toTrack()
        let trackId = optionalInt64("id").map(String.init)
            ?? track?.preview.ref.trackId.rawValue
        guard let trackId else {
            throw MusicLibraryError.invalidResponse(
                description: "Playlist track entry does not contain track id."
            )
        }
        return PlaylistTrackEntry(
            position: position,
            addedAt: optionalString("timestamp"),
            originalIndex: optionalInt64("originalIndex").map(Int.init),
            originalShuffleIndex: optionalInt64("originalShuffleIndex").map(Int.init),
            isRecent: optionalBool("recent"),
            trackRef: TrackRef(
                trackId: TrackId(rawValue: trackId),
                albumId: track?.preview.ref.albumId
            ),
            track: track
        )
    }
}

private let yandexFavouritesPlaylistKind: Int64 = 3

// MARK: - Saved tracks

func parseSavedTracksResult(from value: Any, userId: ProviderUserId) throws -> SavedTracksResult {
    if let sentinel = value as? String {
        if sentinel == "private-library" {
            return .privateLibrary
        }
        throw MusicLibraryError.invalidResponse(
            description: "Unexpected saved tracks sentinel '\(sentinel)'."
        )
    }

    guard let root = value as? [String: Any] else {
        throw MusicLibraryError.invalidResponse(
            description: "Saved tracks result is not a JSON object or string."
        )
    }

    // likes/tracks wraps the list in a "library" key
    let library = root.optionalObject("library") ?? root
    return try .available(library.toSavedTracks())
}

private extension [String: Any] {
    func playlistRole() -> PlaylistRole {
        let kind = optionalInt64("kind")
        return kind == yandexFavouritesPlaylistKind ? .favourites : .regular
    }

    func toSavedTracks() throws -> SavedTracks {
        guard let ownerId = optionalInt64("uid").map(String.init) else {
            throw MusicLibraryError.invalidResponse(
                description: "Saved tracks response does not contain uid."
            )
        }
        // API inconsistency: "revision" vs "revisions"
        let revision = optionalInt64("revision") ?? optionalInt64("revisions")
        let trackItems = optionalObjectArray("tracks") ?? []
        let tracks = try trackItems.enumerated().map { index, element in
            try element.toSavedTrackEntry(position: index)
        }
        return SavedTracks(
            ownerId: ProviderUserId(rawValue: ownerId),
            revision: revision,
            tracks: tracks
        )
    }

    func toSavedTrackEntry(position: Int) throws -> SavedTrackEntry {
        return SavedTrackEntry(
            position: position,
            addedAt: optionalString("timestamp"),
            trackRef: TrackRef(
                trackId: TrackId(rawValue: try requiredStringOrInt("id")),
                albumId: optionalStringOrInt("albumId").map(AlbumId.init(rawValue:))
            ),
            track: try optionalObject("track")?.toTrack().preview
        )
    }
}

// MARK: - Track

extension [String: Any] {
    func toTrack() throws -> Track {
        let trackId = optionalStringOrInt("id")
        guard let trackId else {
            throw MusicLibraryError.invalidResponse(
                description: "Track response does not contain id."
            )
        }

        let albumId = optionalObjectArray("albums")?
            .first?
            .optionalInt64("id")
            .map(String.init)

        let preview = TrackPreview(
            ref: TrackRef(
                trackId: TrackId(rawValue: trackId),
                albumId: albumId.map(AlbumId.init(rawValue:))
            ),
            title: try requiredString("title"),
            artists: optionalObjectArray("artists")?
                .compactMap { try? $0.toArtistPreview() } ?? [],
            durationMs: optionalInt64("durationMs"),
            coverUriTemplate: optionalString("coverUri") ?? optionalString("ogImage"),
            isAvailable: optionalBool("available") ?? true
        )

        return Track(
            preview: preview,
            lyricsAvailable: optionalBool("lyricsAvailable") ?? false,
            isAvailableForPremium: optionalBool("availableForPremiumUsers") ?? false,
            isAvailableWithoutPermission: optionalBool("availableFullWithoutPermission") ?? false
        )
    }

    private func toArtistPreview() throws -> ArtistPreview {
        guard let id = optionalStringOrInt("id"), let name = optionalString("name") else {
            throw MusicLibraryError.invalidResponse(
                description: "Artist entry is missing id or name."
            )
        }
        return ArtistPreview(id: id, name: name)
    }
}

// MARK: - TrackRef → API track id

extension TrackRef {
    func toApiTrackId() -> String {
        albumId.map { "\(trackId.rawValue):\($0.rawValue)" } ?? trackId.rawValue
    }
}

// MARK: - Cover URI

private extension [String: Any] {
    func coverUriTemplate() -> String? {
        let cover = optionalObject("cover")
        switch cover?.optionalString("type") {
        case "pic":
            return cover?.optionalString("uri") ?? optionalString("ogImage")

        case "mosaic":
            let firstItem = cover?.optionalAnyArray("itemsUri")?.first as? String
            return firstItem ?? optionalString("ogImage")

        default:
            return optionalString("ogImage")
        }
    }
}

// MARK: - Playlist visibility

private extension String {
    var toPlaylistVisibility: PlaylistVisibility? {
        switch self {
        case "public": return .public
        case "private": return .private
        default: return nil
        }
    }
}

File: android/services/yandex/src/main/java/com/mplayeraudio/services/yandex/internal/YandexMusicMappers.kt
package com.mplayeraudio.services.yandex.internal

import com.mplayeraudio.core.domain.musiclibrary.AlbumId
import com.mplayeraudio.core.domain.musiclibrary.ArtistPreview
import com.mplayeraudio.core.domain.musiclibrary.MusicLibraryException
import com.mplayeraudio.core.domain.musiclibrary.MusicProviderId
import com.mplayeraudio.core.domain.musiclibrary.MusicServiceAvailability
import com.mplayeraudio.core.domain.musiclibrary.Playlist
import com.mplayeraudio.core.domain.musiclibrary.PlaylistId
import com.mplayeraudio.core.domain.musiclibrary.PlaylistKind
import com.mplayeraudio.core.domain.musiclibrary.PlaylistSummary
import com.mplayeraudio.core.domain.musiclibrary.PlaylistTrackEntry
import com.mplayeraudio.core.domain.musiclibrary.PlaylistUuid
import com.mplayeraudio.core.domain.musiclibrary.PlaylistRole
import com.mplayeraudio.core.domain.musiclibrary.PlaylistVisibility
import com.mplayeraudio.core.domain.musiclibrary.ProviderUserId
import com.mplayeraudio.core.domain.musiclibrary.SavedTrackEntry
import com.mplayeraudio.core.domain.musiclibrary.SavedTracks
import com.mplayeraudio.core.domain.musiclibrary.SavedTracksResult
import com.mplayeraudio.core.domain.musiclibrary.Track
import com.mplayeraudio.core.domain.musiclibrary.TrackId
import com.mplayeraudio.core.domain.musiclibrary.TrackPreview
import com.mplayeraudio.core.domain.musiclibrary.TrackRef
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject

internal fun JsonObject.toAvailability(): MusicServiceAvailability {
    val account = requiredObject("account")
    val permissions = optionalObject("permissions")
    return MusicServiceAvailability(
        isAvailable = account.requiredBoolean("serviceAvailable"),
        region = account.optionalLong("region")?.toInt(),
        permissions = permissions?.optionalArray("values")
            .orEmpty()
            .mapNotNull(JsonElement::stringContent)
            .toSet(),
    )
}

internal fun JsonObject.toCurrentUserId(): ProviderUserId {
    val id = optionalString("id") ?: optionalString("default_uid")
    return id?.let(::ProviderUserId)
        ?: throw MusicLibraryException.InvalidResponse(
            description = "Yandex user info response does not contain user id.",
        )
}

internal fun JsonObject.toPlaylistSummary(): PlaylistSummary {
    val owner = requiredObject("owner")
    val ownerId = owner.optionalLong("uid")?.toString() ?: requiredLong("uid").toString()
    val kind = requiredLong("kind")
    return PlaylistSummary(
        id = PlaylistId(
            ownerId = ProviderUserId(ownerId),
            kind = PlaylistKind(kind),
        ),
        provider = MusicProviderId.YandexMusic,
        playlistUuid = optionalString("playlistUuid")?.let(::PlaylistUuid),
        title = requiredString("title"),
        ownerName = owner.optionalString("name"),
        coverUriTemplate = coverUriTemplate(),
        trackCount = requiredLong("trackCount").toInt(),
        durationMs = optionalLong("durationMs"),
        isAvailable = optionalBoolean("available") ?: true,
        isCollective = optionalBoolean("collective") ?: false,
        visibility = optionalString("visibility")?.toPlaylistVisibility(),
        role = if (kind == YandexFavouritesPlaylistKind) PlaylistRole.Favourites else PlaylistRole.Regular,
    )
}

private const val YandexFavouritesPlaylistKind = 3L

internal fun JsonObject.toPlaylist(): Playlist {
    val tracks = optionalArray("tracks")
        .orEmpty()
        .mapIndexed { index, element -> element.jsonObject.toPlaylistTrackEntry(index) }

    return Playlist(
        summary = toPlaylistSummary(),
        revision = optionalLong("revision"),
        snapshot = optionalLong("snapshot"),
        likesCount = optionalLong("likesCount")?.toInt(),
        tracks = tracks,
    )
}

internal fun JsonElement.toSavedTracksResult(): SavedTracksResult {
    val sentinel = stringContent()
    if (sentinel == "private-library") {
        return SavedTracksResult.PrivateLibrary
    }

    if (this is JsonPrimitive) {
        throw MusicLibraryException.InvalidResponse(
            description = "Unexpected saved tracks sentinel '$content'.",
        )
    }

    val root = jsonObject
    val library = root.optionalObject("library") ?: root
    return SavedTracksResult.Available(value = library.toSavedTracks())
}

internal fun TrackRef.toApiTrackId(): String {
    return albumId?.let { "${trackId.value}:${it.value}" } ?: trackId.value
}

private fun JsonObject.toPlaylistTrackEntry(position: Int): PlaylistTrackEntry {
    val track = optionalObject("track")?.toTrack()
    val trackId = optionalLong("id")?.toString()
        ?: track?.preview?.ref?.trackId?.value
        ?: throw MusicLibraryException.InvalidResponse(
            description = "Playlist track entry does not contain track id.",
        )

    return PlaylistTrackEntry(
        position = position,
        addedAt = optionalString("timestamp"),
        originalIndex = optionalLong("originalIndex")?.toInt(),
        originalShuffleIndex = optionalLong("originalShuffleIndex")?.toInt(),
        isRecent = optionalBoolean("recent"),
        trackRef = TrackRef(
            trackId = TrackId(trackId),
            albumId = track?.preview?.ref?.albumId,
        ),
        track = track,
    )
}

private fun JsonObject.toSavedTracks(): SavedTracks {
    val ownerId = optionalLong("uid")?.toString()
        ?: throw MusicLibraryException.InvalidResponse(
            description = "Saved tracks response does not contain uid.",
        )

    val revision = optionalLong("revision") ?: optionalLong("revisions")
    val tracks = optionalArray("tracks")
        .orEmpty()
        .mapIndexed { index, element -> element.jsonObject.toSavedTrackEntry(index) }

    return SavedTracks(
        ownerId = ProviderUserId(ownerId),
        revision = revision,
        tracks = tracks,
    )
}

private fun JsonObject.toSavedTrackEntry(position: Int): SavedTrackEntry {
    return SavedTrackEntry(
        position = position,
        addedAt = optionalString("timestamp"),
        trackRef = TrackRef(
            trackId = TrackId(requiredStringOrLong("id")),
            albumId = optionalStringOrLong("albumId")?.let(::AlbumId),
        ),
        track = optionalObject("track")?.toTrack()?.preview,
    )
}

internal fun JsonObject.toTrack(): Track {
    val trackId = optionalString("id") ?: optionalLong("id")?.toString()
    val ref = TrackRef(
        trackId = trackId?.let(::TrackId)
            ?: throw MusicLibraryException.InvalidResponse(
                description = "Track response does not contain id.",
            ),
        albumId = optionalArray("albums")
            ?.firstOrNull()
            ?.jsonObject
            ?.optionalLong("id")
            ?.toString()
            ?.let(::AlbumId),
    )

    val preview = TrackPreview(
        ref = ref,
        title = requiredString("title"),
        artists = optionalArray("artists")
            .orEmpty()
            .mapNotNull { artist -> artist.jsonObject.optionalArtistPreview() },
        durationMs = optionalLong("durationMs"),
        coverUriTemplate = optionalString("coverUri") ?: optionalString("ogImage"),
        isAvailable = optionalBoolean("available") ?: true,
    )

    return Track(
        preview = preview,
        lyricsAvailable = optionalBoolean("lyricsAvailable") ?: false,
        isAvailableForPremium = optionalBoolean("availableForPremiumUsers") ?: false,
        isAvailableWithoutPermission = optionalBoolean("availableFullWithoutPermission") ?: false,
    )
}

private fun JsonObject.optionalArtistPreview(): ArtistPreview? {
    val id = optionalString("id") ?: optionalLong("id")?.toString()
    val name = optionalString("name")
    return if (id != null && name != null) {
        ArtistPreview(
            id = id,
            name = name,
        )
    } else {
        null
    }
}
